File: 113. Path Sum II/113. Path Sum II.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func findPaths(root *TreeNode, targetSum int, path []int, result *[][]int) {
    if root == nil {
        return
    }
    
    path = append(path, root.Val)
    
    if root.Left == nil && root.Right == nil && targetSum == root.Val {
        pathCopy := make([]int, len(path))
        copy(pathCopy, path)
        *result = append(*result, pathCopy)
        return
    }
    
    findPaths(root.Left, targetSum - root.Val, path, result)
    findPaths(root.Right, targetSum - root.Val, path, result)
}

func pathSum(root *TreeNode, targetSum int) [][]int {
    result := [][]int{}
    findPaths(root, targetSum, []int{}, &result)
    return result
}

File: 113. Path Sum II/113. Path Sum II.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function pathSum(root: TreeNode | null, targetSum: number): number[][] {
    const result: number[][] = [];
    
    function findPaths(node: TreeNode | null, target: number, path: number[]): void {
        if (!node) return;
        
        path.push(node.val);
        
        if (!node.left && !node.right && target === node.val) {
            result.push([...path]);
        }
        
        findPaths(node.left, target - node.val, path);
        findPaths(node.right, target - node.val, path);
        
        path.pop();
    }
    
    findPaths(root, targetSum, []);
    return result;
}